coding/python_testing.rst

```diff
@@ -42,44 +41,134 @@ The important things to note in this example are:
-`pytest`_ test discovery is much more flexible than that provided by :mod:`unittest`, but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner.
+The `pytest`_ scheme for discovering tests inside Python modules is much more flexible than that provided by :mod:`unittest`, but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner, and it is required that an individual test file can be executed by running it directly with :command:`python`.
 In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
 
+The behavior of `pytest`_ when invoked by :command:`scons` is controlled by the :file:`tests/SConscript` file.
+At its simplest, this file should contain the following to enable testing:
+
+.. code-block:: python
+
+   from lsst.sconsUtils import scripts
+   scripts.BasicSConscript.tests()
+
+When the ``tests`` target is executed the Python test files will be located with a glob for ``*.py`` in the :file:`tests` directory and the tests will be run with a single invocation of :command:`pytest` using the number of subprocesses matching the ``-j`` argument given to :command:`scons`.
+This is the classical method for test discovery but the preferred approach now is for packages to support automatic test discovery using `pytest`_ as described below.
+
+The test file discovery mode can be controlled with two arguments to the :lmeth:`~lsst.sconsUtils.scripts.BasicSConscript.tests` method:
+
+.. code-block:: python
+
+   scripts.BasicSConscript.tests(pyList=[], pySingles=["testSingle.py"])
+
+``pyList`` is used to specify which Python test files to run.
+Here the empty list is interpreted as "allow :command:`pytest` to automatically discover tests", whereas, for backwards compatibility reasons, the default of ``pyList=None`` is interpreted as "allow :lmeth:`~lsst.sconsUtils.scripts.BasicSConscript.tests` to automatically discover tests", as described above.
+For this mode to work, all test files must be named :file:`test_*.py`.
+
+``pySingles`` is an optional argument that can be used for the rare cases where a test must be run standalone and not with other test files in a shared process.
+The tests are still run using :command:`pytest` but executed one at a time without using any multi-process execution.
+Use of this should be extremely rare.
+In the :lmod:`~lsst.base` package one test file is used to confirm that the LSST import code is working; this can only be tested if we know that it hasn't previously been imported as part of another test.
+The other reason, so far, to run a test standalone is for test classes that dynamically generate large amounts of test data in setup.
```

Comment:
"setup" or "setUp"?

Reply:
"the set up phase" (since it's most like `setUpClass`).